meshmode/array_context.py

```diff
@@ -437,77 +437,73 @@ def _get_array_container_ac_arith(ary: ArrayContainerWithArithmetic):
+    is checked for whether it is a subclass of :class:`ArrayContainer`.
+    Note that this classification does not necessarily agree with
+    :class:`is_array_container` (which is not usable here because it operates
+    on instances, not types). :mod:`numpy` object arrays are a particularly
+    important example of this. To document that an attribute *is* a
+    :mod:`numpy` object array, use :class:`NumpyObjectArray` in the
+    attribute type annotation.
+    """
+    from dataclasses import dataclass
+    cls = dataclass(frozen=True)(cls)
 
-    return exec_dict["_generated_serialize_dataclass"]
+    array_fields = [f for f in fields(cls)
+            if issubclass(f.type, _ARRAY_CONTAINER_LIKE)]
+    non_array_fields = [f for f in fields(cls)
+            if not issubclass(f.type, _ARRAY_CONTAINER_LIKE)]
 
+    if not array_fields:
+        raise ValueError(f"'{cls}' must have fields with array container type "
+                "in order to use the dataclass_array_container decorator")
 
-@memoize(key=lambda ary: (type(ary), _deserialize_dataclass_container))
-def _generate_dataclass_deserialization_func(template):
-    from pytools.codegen import remove_common_indentation
-    code = remove_common_indentation("""
-    def _generated_deserialize_dataclass(_template, _iterable):
-        kwargs = dict(_iterable)
-        return type(_template)({})
-    """.format(
-        ", ".join(
-            f"{f.name}=kwargs['{f.name}']"
-            if issubclass(f.type, (ArrayContainer, NumpyObjectArray))
-            else f"{f.name}=_template.{f.name}"
-            for f in fields(template))
-        )
-    )
+    ser_expr = ", ".join(f"({repr(f.name)}, ary.{f.name})" for f in array_fields)
+    template_kwargs = "".join(
+            f"{f.name}=template.{f.name}, " for f in non_array_fields)
 
-    exec_dict = {}
-    exec(compile(code, "<generated code>", "exec"), exec_dict)
-    exec_dict["_MODULE_SOURCE_CODE"] = code
+    from pytools.codegen import remove_common_indentation
+    ser_code = remove_common_indentation(f"""
+        from typing import Any, Iterable, Tuple
+        from meshmode.array_context import serialize_container, deserialize_container
 
-    return exec_dict["_generated_deserialize_dataclass"]
+        @serialize_container.register(cls)
+        def serialize_{cls.__name__}(ary: cls):
+            return ({ser_expr},)
 
+        @deserialize_container.register(cls)
+        def deserialize_{cls.__name__}(
+                template: cls, iterable: Iterable[Tuple[Any, Any]]) -> cls:
+            return cls({template_kwargs}**dict(iterable))
+        """)
 
-@serialize_container.register(DataclassArrayContainer)
-def _serialize_dataclass_container(ary: DataclassArrayContainer):
-    return _generate_dataclass_serialization_func(ary)(ary)
+    # FIXME: The fact that serialization goes to an iterable (and not a dict)
+    # forces repeated iteration over the iterable here (we're having to call dict()).
```

Comment:
What do you think here? The downside would be that we wouldn't allow generators any more, but TBH I wouldn't expect array containers to be sufficiently large for this to matter.

Reply:
Is the suggestion to make the serialization return a dict instead? I'm fine with that.

Reply:
That *was* the suggestion. But using a dict forces double iteration onto the object array to convert from `ndenumerate`, so IMO there's just no winning. I'll remove the FIXME and leave it the way it is. Plus relying on order-dependent `dict`s still makes me feel unwell.

c7cbd75